tests/components/recorder/auto_repairs/test_schema.py

```diff
@@ -0,0 +1,149 @@
+    assert "Schema validation failed" not in caplog.text
+    assert "Detected statistics schema errors" not in caplog.text
+    assert "Database is about to correct DB schema errors" not in caplog.text
+
+
+async def test_validate_db_schema_fix_utf8_issue_good_schema(
+    async_setup_recorder_instance: RecorderInstanceGenerator,
+    hass: HomeAssistant,
+    recorder_db_url: str,
+    caplog: pytest.LogCaptureFixture,
+) -> None:
+    """Test validating DB schema with MySQL when the schema is correct."""
+    if not recorder_db_url.startswith("mysql://"):
+        # This problem only happens on MySQL
+        return
+    await async_setup_recorder_instance(hass)
+    await async_wait_recording_done(hass)
+    instance = get_instance(hass)
+    schema_errors = await instance.async_add_executor_job(
+        validate_table_schema_supports_utf8, instance, States, ("state",)
+    )
+    assert schema_errors == set()
+
+
+async def test_validate_db_schema_fix_utf8_issue_with_broken_schema(
+    async_setup_recorder_instance: RecorderInstanceGenerator,
+    hass: HomeAssistant,
+    recorder_db_url: str,
+    caplog: pytest.LogCaptureFixture,
+) -> None:
+    """Test validating DB schema with MySQL when the schema is broken."""
+    if not recorder_db_url.startswith("mysql://"):
+        # This problem only happens on MySQL
+        return
+    await async_setup_recorder_instance(hass)
+    await async_wait_recording_done(hass)
+    instance = get_instance(hass)
+    session_maker = instance.get_session
+
+    def _break_states_schema():
+        with session_scope(session=session_maker()) as session:
+            session.execute(
+                text(
+                    "ALTER TABLE states MODIFY state VARCHAR(255) "
+                    "CHARACTER SET ascii COLLATE ascii_general_ci, "
+                    "LOCK=EXCLUSIVE;"
+                )
+            )
+
+    await instance.async_add_executor_job(_break_states_schema)
+    schema_errors = await instance.async_add_executor_job(
+        validate_table_schema_supports_utf8, instance, States, ("state",)
+    )
+    assert schema_errors == {"states.4-byte UTF-8"}
+
+
+async def test_validate_db_schema_precision_good_schema(
+    async_setup_recorder_instance: RecorderInstanceGenerator,
+    hass: HomeAssistant,
+    recorder_db_url: str,
+    caplog: pytest.LogCaptureFixture,
+) -> None:
+    """Test validating DB schema when the schema is correct."""
+    if not recorder_db_url.startswith(("mysql://", "postgresql://")):
+        # This problem only happens on MySQL and PostgreSQL
+        return
+    await async_setup_recorder_instance(hass)
+    await async_wait_recording_done(hass)
+    instance = get_instance(hass)
+    schema_errors = await instance.async_add_executor_job(
+        validate_db_schema_precision,
+        instance,
+        States,
+    )
+    assert schema_errors == set()
+
+
+async def test_validate_db_schema_precision_with_broken_schema(
```

Comment:
We could make a test that actually corrects the schema and than makes sure there are no validation errors

Reply:
Maybe better to not make this PR any bigger and expand the test coverage when we add events